src/io/webcam.rs

```diff
@@ -0,0 +1,247 @@
+    ///
+    /// A WebcamCapture object
+    fn new(camera_id: usize, size: Option<ImageSize>) -> Result<Self> {
+        gst::init()?;
+
+        // create a pipeline specified by the camera id and size
+        let pipeline_str = Self::gst_pipeline_string(camera_id, size);
+        let pipeline = gst::parse::launch(&pipeline_str)?
+            .downcast::<gst::Pipeline>()
+            .map_err(|_| anyhow::anyhow!("Failed to downcast pipeline"))?;
+
+        let appsink = pipeline
+            .by_name("sink")
+            .ok_or_else(|| anyhow::anyhow!("Failed to get sink"))?
+            .dynamic_cast::<gst_app::AppSink>()
+            .map_err(|_| anyhow::anyhow!("Failed to cast to AppSink"))?;
+
+        let (tx, rx) = std::sync::mpsc::channel();
+
+        appsink.set_callbacks(
+            gst_app::AppSinkCallbacks::builder()
+                .new_sample(move |sink| match Self::extract_image_frame(sink) {
+                    Ok(frame) => {
+                        if let Err(e) = tx.send(frame) {
+                            println!("Error sending frame: {}", e);
+                            Err(gst::FlowError::Error)
+                        } else {
+                            Ok(gst::FlowSuccess::Ok)
+                        }
+                    }
+                    Err(e) => {
+                        println!("Error extracting image frame: {}", e);
+                        Err(gst::FlowError::Error)
+                    }
+                })
+                .build(),
+        );
+
+        pipeline.set_state(gst::State::Playing)?;
+
+        // Event loop or processing here
+        // Set pipeline state to Null when done
+
+        let bus = pipeline
+            .bus()
+            .ok_or_else(|| anyhow::anyhow!("Failed to get bus"))?;
+
+        let handle = std::thread::spawn(move || {
+            for msg in bus.iter_timed(gst::ClockTime::NONE) {
+                use gst::MessageView;
+                match msg.view() {
+                    MessageView::Eos(..) => break,
+                    MessageView::Error(err) => {
+                        eprintln!(
+                            "Error from {:?}: {}",
+                            msg.src().map(|s| s.path_string()),
+                            err.error(),
+                            //err.error_code()
+                        );
+                        break;
+                    }
+                    _ => (),
+                }
+            }
+        });
+        Ok(Self {
+            pipeline,
+            receiver: rx,
+            handle: Some(handle),
+        })
+    }
+
+    /// Runs the webcam capture object and grabs frames from the camera
+    ///
+    /// # Arguments
+    ///
+    /// * `f` - A function that takes an image frame
+    pub fn run<F>(&self, f: F) -> Result<()>
+    where
+        F: Fn(Image<u8, 3>),
+    {
+        let receiver = &self.receiver;
+        while let Ok(img) = receiver.recv() {
+            f(img);
+        }
```

Comment:
looks like there's no way for `f` to stop this right now, that seems problematic!